### PR TITLE
feat: edit/delete/reply for emoji/sticker/audio

### LINKED ISSUE #13246: Add missing actions (edit/delete/reply) to certain types of messages (emoji/sticker/audio)
# Feature Issue

Currently, certain types of chat messages (emoji/sticker/audio) are missing some actions (edit/delete/reply) that can be applicable to those types of messages but not yet implemented. Therefore, we should add those missing actions (see description below). 

### User Story

As a user, I want to have possibility to **edit/delete** _**emoji**_ messages that are sent by me in 1-1/group chats. 
As a user, I want to have possibility to **reply** to _**audiomessages**_ that are received/sent by me in 1-1/group chats. 
As a user, I want to have possibility to **reply** to **_sticker_** messages that are received/sent by me and **delete** this type of messages sent by me in 1-1/group/public chats. 

### Description

In 1-1/group chats:
1. **_emoji_**: add **edit/delete** actions (for messages sent by user)
2. _**audiomessage**_: add **reply** action
3. **_sticker_**: add **reply** action. For messages sent by user - add **delete** action.

In public chat:
1. **_sticker_**: add **reply** action. For messages sent by user - add **delete** action.

### PR DESCRIPTION
**Pull request by OmarBasem**
#### You can find continuation of this PR at that new [PR](https://github.com/status-im/status-mobile/pull/13808)

Fixes #13246 

### Summary

Add missing actions (edit/delete/reply) to certain types of messages (emoji/sticker/audio)
To implement the needed functionalities there are 2 pull requests to this repo `status-mobile` and [status-go](https://github.com/status-im/status-go/pull/2771).

1. Editing emoji messages: Firstly, I had to allow the emoji content type to be edited from status-go. Secondly, when editing an emoji message, I had to check whether it is still an emoji message or is now a text message, and send the contentType. Also, in status-go I added a "ContentType" field to the "EditMessage" struct and updated the content type when a message is edited. Regarding UI, I had to make sure a white space is added at the end of an emoji message when edited (similar to text messages) to avoid overlapping with the "edited" text and icon.

2. Deleting emoji messages: added delete option on-long-press.

3. Audio message reply: In status-go, I added an AudioLocalURL field to the QuotedMessage struct. When an audio message is replied to, the AudioLocalURL field will be assigned to the QuotedMessage. In status-mobile, inside the quoted-message method, I check if the quoted message contains audio and if so, will render the audio message.

4. Sticker message reply: Initially, I attempted to do it in a similar way to replying to audio, but it didn't work as the sticker url is created using the sticker hash and not the message ID. Currently, the "responseTo" field includes the messageID only. So, what I did is simply get the sticker url from the message itself using the messageID inside the quoted-message method in status-mobile. I added a `HasSticker` field to the QuotedMessage struct in status-go to know if it is a sticker. Additionally, if the sticker is deleted, a message will be shown to the user stating "This message has been deleted!", which I think can also be implemented for replying to other content types.

5. Sticker message delete: added delete option on-long-press.

Sticker reply and delete also work in public chats.

### Review Notes
This PR depends on [that status-go PR](https://github.com/status-im/status-go/pull/2771) for backend functionality: 87db7d953e8fc748baa1fc56a7adffd8fada3055

#### Platforms

- Android
- iOS

status: ready

### CODE REVIEW

**Comment by Parveshdhull** [comment]
Hi @OmarBasem, Thank you very much for PR.

Looks like linting and tests are failed.

You can run `make lint` to see the linting errors and fix them manually or run `make lint-fix`.
And you can checkout errors with tests with the help of `make test`.

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 1]
```diff
@@ -86,28 +86,51 @@
...
+                     ;; Disabling images for public-chats
+                     (not public?))
+              [fast-image/fast-image
+               {:style  {:width            56
+                         :height           56
+                         :background-color :black
+                         :border-radius    4}
+                :source {:uri image}}]
+
+              (if (and audio
```
Comment (Parveshdhull) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: Better to use [cond](https://clojuredocs.org/clojure.core/cond), instead of multiple `if`

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 2]
```diff
@@ -86,28 +86,51 @@
...
+                       ;; Disabling audio for public-chats
+                       (not public?))
+                [react/view {:style (style/message-view message) :accessibility-label :audio-message}
+                 [react/view {:style (style/message-view-content)}
+                  [react/view {:style (style/style-message-text outgoing)}
+                   [message.audio/message-content message]
+                   [message-status message]]]]
+
+                (if sticker
+                  (if (def repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id))
```
Comment (Parveshdhull) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: - Please move subscription to letsubs
- Also `(def ...)` returns nil, so `(if (def ...))` will be always false
- @(re-frame/subscribe ...) will return message object and by surrounding this with (), you are using this as a function, wouldn't that will throw error?

**Comment by OmarBasem** [comment]
Hi @Parveshdhull, thank you for your comments!

- I fixed the linting errors using `make lint-fix`.
- I double-checked running the tests using `make test` and all the tests are passing.
- I updated the code to use `cond` in the mentioned snippet.
- Moved `repliedMsg` subscription to `letsubs`

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 2 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: - The line `(if (def repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id))` will be true if the repliedMsg exists (has not been deleted), otherwise will be false. I have refactored it to `letsubs` anyway.
- `(@(re-frame/subscribe [:chats/chat-messages chat-id]) id)` is used to get a particular message (the replied-to message), hence the parentheses are needed. Actually removing the parentheses would throw an error.

**Comment by flexsurfer** [comment]
hey @OmarBasem thank you for the PR, its little bit difficult to review PR, could you please revert formatting , thanks

**Comment by OmarBasem** [comment]
> hey @OmarBasem thank you for the PR, its little bit difficult to review PR, could you please revert formatting , thanks

Hey @flexsurfer, sorry for that, my IDE couldn't format clojurescript properly. I have fixed the formatting.

**Comment by OmarBasem** [comment]
Hey @flexsurfer, I am not sure here why the commit cannot be built? It has already been built previously before reformatting. I did not add or remove any code, but now it is not building. I checked lint and tests - they are all passing.

**Comment by Parveshdhull** [comment]
> Hey @flexsurfer, I am not sure here why the commit cannot be built? It has already been built previously before reformatting. I did not add or remove any code, but now it is not building. I checked lint and tests - they are all passing.

Hi, you need to rebase your branch with develop
image

**Comment by OmarBasem** [comment]
Hi @Parveshdhull, I just rebased, and it still cannot be built.

**Comment by Parveshdhull** [comment]
> Hi @Parveshdhull, I just rebased, and it still cannot be built.

It is not properly rebased yet, error reproduction steps
```
git clone https://github.com/OmarBasem/status-mobile && cd status-mobile
git remote add status https://github.com/status-im/status-mobile
git fetch status
git rebase status/develop
```

there are still conflicts in `ios/Podfile.lock`

**Comment by OmarBasem** [comment]
Hey @Parveshdhull @flexsurfer, I think the PR is ready to be reviewed?

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 3]
```diff
@@ -87,8 +87,9 @@
         timestamp-str]])))
 
 (defview quoted-message
-  [_ {:keys [from parsed-text image]} outgoing current-public-key public? pinned]
-  (letsubs [contact-name [:contacts/contact-name-by-identity from]]
+  [_ {:keys [from parsed-text image audio sticker id] :as message} outgoing current-public-key public? pinned chat-id]
+  (letsubs [contact-name [:contacts/contact-name-by-identity from]
+            repliedMsg (@(re-frame/subscribe [:chats/chat-messages chat-id]) id)] ;; Msg replied to
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: it probably makes sense to stop using `defview`/`letsubs` here

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 4]
```diff
@@ -97,17 +98,33 @@
...
+          [message.audio/message-content message]
+          [message-status message]]]]
+
+       sticker
+       (if repliedMsg
+         [fast-image/fast-image {:style  {:margin 4 :width 56 :height 56}
+           ;; Get the sticker url of the message replied to
+                                 :source {:uri (((repliedMsg :content) :sticker) :url)}}]
+          ;; Let the user know if the message was deleted
+         [react/text {:style (style/quoted-message-text (and outgoing (not pinned)))}
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: Could you please align comments?

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 5]
```diff
@@ -467,35 +486,37 @@
...
               [react/view {:on-layout
-                           #(when (and (> (.-nativeEvent.layout.height ^js %) max-message-height-px)
-                                       (not @collapsible?)
-                                       (not outgoing)
-                                       (not modal))
+                           #(when
+                             (and (> (.-nativeEvent.layout.height ^js %) max-message-height-px)
+                                  (not @collapsible?)
+                                  (not outgoing)
+                                  (not modal))
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: unnecessary change

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 6]
```diff
@@ -467,35 +486,37 @@
...
-                 [quoted-message response-to (:quoted-message message) outgoing current-public-key public? pinned])
+                 [quoted-message response-to (:quoted-message message) outgoing current-public-key public? pinned chat-id])
                [render-parsed-text-with-message-status message (:parsed-text content)]])
             (when-not @collapsible? [message-status message])
             (when (and @collapsible? (not modal))
               (if @collapsed?
-                (let [color (if pinned colors/pin-background (if mentioned colors/mentioned-background colors/blue-light))]
+                (let [color (if pinned
+                              colors/pin-background
+                              (if mentioned colors/mentioned-background colors/blue-light))]
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: unnecessary change

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 7]
```diff
@@ -467,35 +486,37 @@
...
-                   [react/linear-gradient {:colors [(str color "00") color]
-                                           :start  {:x 0 :y 0} :end {:x 0 :y 0.9}}
-                    [react/view {:height         72 :align-self :center :justify-content :flex-end
-                                 :padding-bottom 10}
+                   {:on-press #(swap! collapsed? not) :style {:position :absolute :bottom 0 :left 0 :right 0 :height 72}}
+                   [react/linear-gradient
+                    {:colors [(str color "00") color]
+                     :start  {:x 0 :y 0}
+                     :end    {:x 0 :y 0.9}}
+                    [react/view {:height 72 :align-self :center :justify-content :flex-end :padding-bottom  10}
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: looks like unnecessary change

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 8]
```diff
@@ -467,35 +486,37 @@
...
                      [react/view (style/collapse-button)
                       [icons/icon :main-icons/dropdown
                        {:color colors/white}]]]]])
-                [react/touchable-highlight {:on-press #(swap! collapsed? not)
-                                            :style    {:align-self :center :margin 5}}
+                [react/touchable-highlight
+                 {:on-press #(swap! collapsed? not) :style {:align-self :center :margin 5}}
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: unnecessary change

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 9]
```diff
@@ -520,65 +541,97 @@
...
-                [quoted-message response-to (:quoted-message message) outgoing current-public-key public? pinned])
-              [react/text {:style (style/emoji-message message)}
-               (:text content)]]
-             [message-status message]]]]]
-         reaction-picker]))))
+    (fn
+      [{:keys [content current-public-key outgoing edit-enabled public?
+               pinned in-popover? message-pin-enabled content-type edited-at] :as   message}
+       {:keys [on-long-press modal] :as   reaction-picker}]
+               ;; Makes sure to render a text-messsage and not an emoji-message if it has been edited with text
```
Comment (rasom) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: align comment

File: ios/Podfile.lock [hunk 10]
```diff
@@ -635,7 +635,7 @@ SPEC CHECKSUMS:
   FBLazyVector: 3bb422f41b18121b71783a905c10e58606f7dc3e
   FBReactNativeSpec: f2c97f2529dd79c083355182cc158c9f98f4bd6e
   Folly: b73c3869541e86821df3c387eb0af5f65addfab4
-  glog: 36ce0530c6d2c3a5a4326885ef4069564887a1db
+  glog: 64a900d72fb14cd1b2cdf3b717a5555ceb889d49
```
Comment (flexsurfer) [inline ios/Podfile.lock]: could you please revert this file

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 3 above)
Comment (flexsurfer) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: and better to use get `(get @(re-frame/subscribe [:chats/chat-messages chat-id]) id)`

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 11]
```diff
@@ -608,27 +661,37 @@
...
-                                     :label    (i18n/label :t/delete)
-                                     :id       :delete}]
-                                   [])))
+           (fn []
+             (on-long-press
+              (concat
+               [{:on-press #(re-frame/dispatch [:chat.ui/reply-to-message message])
+                 :id       :reply
+                 :label    (i18n/label :t/message-reply)}]
+               (if (and outgoing config/delete-message-enabled?)
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: better to use `when` here

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 12]
```diff
@@ -608,27 +661,37 @@
...
+                   :id       :delete}]
+                 []))))
            :on-press (fn []
                        (reset! show-timestamp? true))})
         [react/view (style/message-view-wrapper (:outgoing message))
          [message-timestamp message show-timestamp?]
          [react/view {:style (style/message-view message) :accessibility-label :audio-message}
           [react/view {:style (style/message-view-content)}
-           [message.audio/message-content message] [message-status message]]]]]
+           [react/view {:style (style/style-message-text outgoing)}
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: why extra view ? its better to have less views as possible

File: src/status_im/chat/models/input.cljs [hunk 13]
```diff
@@ -252,14 +252,18 @@
...
+   {::json-rpc/call [{:method      "wakuext_editMessage"
+                      :params      [{:id           message-id
+                                     :text         text
+                                     :content-type (if (message-content/emoji-only-content? {:text text :response-to quoted-message})
+                                                     constants/content-type-emoji
+                                                     constants/content-type-text)}]
                       :js-response true
-                      :on-error #(log/error "failed to edit message " %)
-                      :on-success #(re-frame/dispatch [:sanitize-messages-and-process-response %])}]}
+                      :on-error    #(log/error "failed to edit message " %)
```
Comment (J-Son89) [inline src/status_im/chat/models/input.cljs]: any reason you added the whitespace? consider reverting if not.

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 3 above)
Comment (J-Son89) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: I think `repliedMsg`is not following conventions. It should probably be `replied-msg`, and consider using `replied-message` instead as the character saving adds no value imo.

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 5 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: My IDE is doing unnecessary changes to clojurescript formatting. Can you please tell me what IDE are you using?

Comment (J-Son89) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: are you running `make lint-fix` before commiting?

I'm using vscode but not sure if it's the best option for the tech stack.

File: src/status_im/ui/screens/chat/message/message_old.cljs [hunk 14]
```diff
@@ -97,17 +98,33 @@
...
-                               :source {:uri image}}]
-       [react/text {:style           (style/quoted-message-text (and outgoing (not pinned)))
-                    :number-of-lines 5}
-        (components.reply/get-quoted-text-with-mentions parsed-text)])]))
+     (cond
+       (and image (not public?)) ;; Disabling images for public-chats
+       [fast-image/fast-image
+        {:style  {:width            56
+                  :height           56
+                  :background-color :black
```
Comment (briansztamfater) [inline src/status_im/ui/screens/chat/message/message_old.cljs]: Maybe you can use a quo color constant instead of `:black`

**Comment by OmarBasem** [comment]
To fix all the unnecessary formatting changes caused by the IDE, I had to open a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808) @flexsurfer @rasom @J-Son89 @briansztamfater @Parveshdhull

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 3 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: @flexsurfer but I need to have the replied-message in a variable to use later on in the `if` condition, and when rendering the reply view. How can I do that with `(get @(re-frame/subscribe [:chats/chat-messages chat-id]) id)` ?

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 4 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: Fixed it

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 5 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: Yes I am running `make lint-fix`. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808)

File: ios/Podfile.lock (hunk 10 above)
Comment (OmarBasem) [reply ios/Podfile.lock]: Fixed it

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 11 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: Fixed it

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 12 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: That's not my code. This is showing as a change due to formatting changes by the IDE. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808).

File: src/status_im/ui/screens/chat/message/message_old.cljs (hunk 14 above)
Comment (OmarBasem) [reply src/status_im/ui/screens/chat/message/message_old.cljs]: That's not my code. This is showing as a change due to formatting changes by the IDE. To fix all the unnecessary formatting changes caused by the IDE, I opened a new [PR](https://github.com/status-im/status-mobile/pull/13808). Please follow on the new [PR](https://github.com/status-im/status-mobile/pull/13808).
